xdsl/interpreter.py

```diff
@@ -0,0 +1,273 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import (IO, Any, Callable, Generator, Iterable, TypeAlias, TypeVar,
+                    ParamSpec)
+
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import OperationInvT, SSAValue, Operation
+from xdsl.utils.exceptions import InterpretationError
+
+
+@dataclass
+class InterpreterFunctions:
+    """
+    A class to hold the Python implementations for Operations. Users should
+    subclass this class, and define the function to run during interpretation.
+    For example:
+
+    ``` python
+    @register_impls
+    class ArithFunctions(InterpreterFunctions):
+
+        @impl(arith.Addi)
+        def run_addi(self, interpreter: Interpreter, op: arith.Addi, 
+                     args: tuple[Any, ...]) -> tuple[Any, ...]:
+            lhs, rhs = args
+            return lhs + rhs, 
+    ```
+
+    The interpreter will take care of fetching the Python values associated with
+    the operand SSAValues, and setting the return values to the appropriate
+    OpResults.
+
+    To override the definition of an operation implementation, subclass the
+    class to override, and redefine the function, annotating it with @impl
+
+    ``` python
+    @register_impls
+    class DebugArithFunctions(ArithFunctions):
+
+        @impl(arith.Addi)
+        def run_addi(self, interpreter: Interpreter, op: arith.Addi, 
+                     args: tuple[Any, ...]) -> tuple[Any, ...]:
+            lhs, rhs = args
+            print(lhs, rhs, lhs + rhs)
+            return lhs + rhs, 
+    ```
+    """
+
+    @classmethod
```

Comment:
It can't be abstract, unfortunately, as Python doesn't detect that it's being implemented in the wrapper. I personally like it as a method, both because as a mental model it doesn't really represent something that the class "owns", and because users of the class shouldn't be calling it anyway.